server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -123,7 +124,17 @@ public ClusterState execute(BatchExecutionContext<JoinTask> batchExecutionContex
             newState = ClusterState.builder(initialState);
         } else {
             logger.trace("processing node joins, but we are not the master. current master: {}", currentNodes.getMasterNode());
-            throw new NotMasterException("Node [" + currentNodes.getLocalNode() + "] not master for join request");
+            throw new NotMasterException(
+                "Node ["
+                    + currentNodes.getLocalNode()
+                    + "] not master for join request. Current known master: ["
+                    + currentNodes.getMasterNode()
+                    + "], current term: ["
+                    + term
+                    + "], current local time: [ "
+                    + TimeValue.timeValueMillis(System.currentTimeMillis())
+                    + "]"
```

Comment:
nit: could we use `Strings.format` rather than all this concatenation?

Reply:
Done